File: apps/desktop/src/lib/api-client.ts
import { getClient } from "@/lib/viben";

/**
 * Options for making an API request
 */
export interface ApiRequestOptions {
  /** HTTP method */
  method: "GET" | "POST" | "PUT" | "DELETE" | "PATCH";
  /** API endpoint path (e.g., "/api/mcp") */
  endpoint: string;
  /** Optional request body (will be JSON serialized) */
  body?: unknown;
  /** Optional Bearer token for authentication */
  authToken?: string;
}

/**
 * Standard paginated response from the platform API
 */
export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

// Default platform URL
const PLATFORM_URL = "https://viben-web.vercel.app";

// Store API base URL in memory (localStorage fallback for persistence)
let apiBaseUrl = PLATFORM_URL;

// Try to restore from localStorage
try {
  const stored = localStorage.getItem("viben-api-base-url");
  if (stored) {
    apiBaseUrl = stored;
  }
} catch {
  // localStorage may not be available
}

/**
 * Make an API request to the platform
 *
 * @param options - Request configuration
 * @returns Promise resolving to the response data
 * @throws Error if the request fails
 *
 * @example
 * ```ts
 * // GET request
 * const data = await apiRequest<McpPackage[]>({
 *   method: "GET",
 *   endpoint: "/api/mcp",
 * });
 *
 * // POST request with auth
 * const result = await apiRequest<CreateResult>({
 *   method: "POST",
 *   endpoint: "/api/mcp",
 *   body: { name: "my-package" },
 *   authToken: "xxx",
 * });
 * ```
 */
export async function apiRequest<T>(options: ApiRequestOptions): Promise<T> {
  const url = `${apiBaseUrl}${options.endpoint}`;
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
  };

  if (options.authToken) {
    headers["Authorization"] = `Bearer ${options.authToken}`;
  }

  const response = await fetch(url, {
    method: options.method,
    headers,
    body: options.body ? JSON.stringify(options.body) : undefined,
  });

  if (!response.ok) {
    let errorMessage = `HTTP ${response.status}`;
    try {
      const errorBody = await response.json();
      errorMessage = errorBody.error || errorMessage;
    } catch {
      // Unable to parse error body
    }
    throw new Error(errorMessage);
  }

  return response.json();
}

/**
 * Get the current API base URL
 *
 * @returns Promise resolving to the base URL string
 */
export async function getApiBaseUrl(): Promise<string> {
  return apiBaseUrl;
}

/**
 * Set the API base URL
 *
 * @param url - New base URL (e.g., "https://viben-web.vercel.app")
 * @throws Error if URL format is invalid
 */
export async function setApiBaseUrl(url: string): Promise<void> {
  // Validate URL format
  try {
    new URL(url);
  } catch {
    throw new Error("Invalid URL format");
  }

  apiBaseUrl = url.replace(/\/$/, ""); // Remove trailing slash

  // Persist to localStorage
  try {
    localStorage.setItem("viben-api-base-url", apiBaseUrl);
  } catch {
    // localStorage may not be available
  }
}

/**
 * Reset API base URL to default
 */
export function resetApiBaseUrl(): void {
  apiBaseUrl = PLATFORM_URL;
  try {
    localStorage.removeItem("viben-api-base-url");
  } catch {
    // localStorage may not be available
  }
}

// Re-export getClient for convenience
export { getClient };

File: apps/desktop/src-tauri/src/commands/api_client.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Mutex;
use tauri::State;

/// Default Platform base URL
const DEFAULT_BASE_URL: &str = "https://viben-web.vercel.app";

/// Managed state for API client configuration
pub struct ApiClientState {
    pub base_url: Mutex<String>,
}

impl Default for ApiClientState {
    fn default() -> Self {
        Self {
            base_url: Mutex::new(DEFAULT_BASE_URL.to_string()),
        }
    }
}

/// Generic API response wrapper
#[derive(Debug, Serialize, Deserialize)]
pub struct ApiResponse<T> {
    pub data: T,
    pub error: Option<String>,
}

/// Make an HTTP request to the platform API
///
/// # Arguments
/// * `method` - HTTP method (GET, POST, PUT, DELETE, PATCH)
/// * `endpoint` - API endpoint path (e.g., "/api/mcp")
/// * `body` - Optional JSON body for POST/PUT/PATCH requests
/// * `auth_token` - Optional Bearer token for authentication
/// * `state` - API client state with base URL
///
/// # Returns
/// JSON response value or error string
#[tauri::command]
pub async fn api_request(
    method: String,
    endpoint: String,
    body: Option<Value>,
    auth_token: Option<String>,
    state: State<'_, ApiClientState>,
) -> Result<Value, String> {
    let base_url = state.base_url.lock().unwrap().clone();
    let url = format!("{}{}", base_url, endpoint);

    let client = reqwest::Client::new();
    let mut request = match method.to_uppercase().as_str() {
        "GET" => client.get(&url),
        "POST" => client.post(&url),
        "PUT" => client.put(&url),
        "DELETE" => client.delete(&url),
        "PATCH" => client.patch(&url),
        _ => return Err(format!("Unsupported HTTP method: {}", method)),
    };

    // Set JSON content type
    request = request.header("Content-Type", "application/json");

    // Add Authorization header if token provided
    if let Some(token) = auth_token {
        request = request.header("Authorization", format!("Bearer {}", token));
    }

    // Add request body for methods that support it
    if let Some(data) = body {
        request = request.json(&data);
    }

    // Execute request
    let response = request
        .send()
        .await
        .map_err(|e| format!("Network request failed: {}", e))?;

    // Check for HTTP errors
    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(format!("HTTP {} - {}", status.as_u16(), error_text));
    }

    // Parse JSON response
    let json: Value = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse JSON response: {}", e))?;

    Ok(json)
}

/// Get the current API base URL
#[tauri::command]
pub fn get_api_base_url(state: State<'_, ApiClientState>) -> String {
    state.base_url.lock().unwrap().clone()
}

/// Set the API base URL
///
/// # Arguments
/// * `url` - New base URL (e.g., "https://viben-web.vercel.app")
#[tauri::command]
pub fn set_api_base_url(url: String, state: State<'_, ApiClientState>) -> Result<(), String> {
    // Validate URL format
    if !url.starts_with("http://") && !url.starts_with("https://") {
        return Err("Invalid URL: must start with http:// or https://".to_string());
    }

    // Remove trailing slash if present
    let normalized_url = url.trim_end_matches('/').to_string();

    let mut base_url = state.base_url.lock().unwrap();
    *base_url = normalized_url;
    Ok(())
}
